src/profile-logic/profile-data.js

```diff
@@ -2549,39 +2548,29 @@ export function filterToDeallocationsMemory(
     const bytes = nativeAllocations.weight[allocationIndex];
     const memoryAddress = nativeAllocations.memoryAddress[allocationIndex];
     if (bytes >= 0) {
-      // Handle the allocation.
+      // This is an allocation.
 
       // Provide a map back to this index.
-      memoryAddressToAllocation.set(memoryAddress, allocationIndex);
-      stackOfOriginalAllocation[allocationIndex] = null;
-    } else {
-      // Lookup the previous allocation.
-      const previousAllocationIndex = memoryAddressToAllocation.get(
-        memoryAddress
-      );
-      if (previousAllocationIndex === undefined) {
-        stackOfOriginalAllocation[allocationIndex] = null;
-      } else {
-        // This deallocation matches a previous allocation. Remove the allocation.
-        stackOfOriginalAllocation[allocationIndex] = previousAllocationIndex;
-        // There is a match, so delete this old association.
-        memoryAddressToAllocation.delete(memoryAddress);
-      }
+      memoryAddressToAllocationSite[memoryAddress] =
+        nativeAllocations.stack[allocationIndex];
+      continue;
     }
-  }
 
-  const newDeallocations = getEmptyBalancedNativeAllocationsTable();
-  for (let i = 0; i < nativeAllocations.length; i++) {
-    const duration = nativeAllocations.weight[i];
-    const stackIndex = stackOfOriginalAllocation[i];
-    if (stackIndex !== null) {
-      newDeallocations.time.push(nativeAllocations.time[i]);
-      newDeallocations.stack.push(stackIndex);
-      newDeallocations.weight.push(duration);
-      newDeallocations.memoryAddress.push(nativeAllocations.memoryAddress[i]);
-      newDeallocations.threadId.push(nativeAllocations.threadId[i]);
-      newDeallocations.length++;
+    // This is a deallocation.
+    // Lookup the previous allocation.
+    const allocationStackIndex = memoryAddressToAllocationSite[memoryAddress];
+    if (allocationStackIndex === undefined) {
+      // This deallocation doesn't match an allocation. Let's bail out.
+      continue;
     }
+
+    // This deallocation matches a previous allocation.
+    newDeallocations.time.push(nativeAllocations.time[allocationIndex]);
+    newDeallocations.stack.push(allocationStackIndex);
+    newDeallocations.weight.push(bytes);
+    newDeallocations.memoryAddress.push(memoryAddress);
+    newDeallocations.threadId.push(nativeAllocations.threadId[allocationIndex]);
+    newDeallocations.length++;
```

Comment:
The algorithm was very wrong: we were adding the allocationIndex as the stackIndex value!

I believe the new implementation is simpler and faster too (there's only 1 loop), and I use a sparse array instead of a Map.

Reply:
Praise: Thanks for the correctness fixes and the simplification.

Thought: I read through the new implementation and it looks correct for what it's trying to accomplish. I didn't verify the bad behavior of the old algorithm.